language/move-compiler/tests/move_check/typing/lambda.move

```diff
@@ -0,0 +1,100 @@
+module 0x8675309::M {
+    use 0x1::XVector;
+
+    public macro fun foreach<T>(v: &vector<T>, action: |&T|) { // expected to be not implemented
+        let i = 0;
+        while (i < XVector::length(v)) {
+            action(XVector::borrow(v, i));
+            i = i + 1;
+        }
+    }
+
+    public macro fun reduce<R, T>(v: vector<T>, accu: R, reducer: |T, R|R): R {
+        while (!XVector::is_empty(&v)) {
+            accu = reducer(XVector::pop_back(&mut v), accu);
+        };
+        accu
+    }
+
+
+    public fun correct_foreach() {
+        let v = vector[1, 2, 3];
+        let sum = 0;
+        foreach!(&v, |e| sum = sum + *e) // expected to be not implemented
+    }
+
+    public fun correct_reduce(): u64 {
+        let v = vector[1, 2, 3];
+        reduce!(v, 0, |t, r| t + r)
+    }
+
+    public fun corrected_nested() {
+        let v = vector[vector[1,2], vector[3]];
+        let sum = 0;
+        foreach!(&v, |e| sum = sum + reduce!(*e, 0, |t, r| t + r));
```

Comment:
I might be nitpicking here, but the lack of separator between the input and output is slightly jarring, is there no way to provide a ":" syntax here as well in the macro action?

Reply:
Do you mean the lambda or the function type signature? The lambda is obviously inspired by Rust, I feel its fine.

The signature is a different problem. As we do not use `->` but `:` for return types, we would go strange looking stuff like
`fun foo(p: |int|:bool)` (notice the consecutive colons). I tried to align the current syntax with lambda, and once used to it, I like it ;)